flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io;
+
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Interface defining couple of essential methods for asynchronous and non blocking data polling.
+ */
+public interface AsyncDataInput<T> {
+	CompletableFuture<?> AVAILABLE = CompletableFuture.completedFuture(null);
+
+	/**
+	 * Poll the next element.
+	 *
+	 * @return If the {@link T} is an {@code Optional}:
+	 * 1. return {@code Optional.of(element)} - next returned element. {@code Optional.empty()} will
+	 * be returned if there is no data to return or if {@link #isFinished()} returns true.
+	 * otherwise:
+	 * 2. next element or null if there is no data to return or if {@link #isFinished()} returns true.
+	 */
+	T pollNext() throws Exception;
```

Comment:
Exception is not mentioned in the doc, even if it is rather clear what that is about.

Reply:
Right, you cannot add a lot of information about that, reason is more that you can see declared exceptions in the docs if you add them. Just a visibility thing.